texastribune/donations
# Add pre-commit JS formatting; integrate ESLint into CI

#### What's this PR do?
+ Adds JS code formatting via Prettier, which is done automatically pre-commit
+ Makes a CircleCI build fail if linting via ESLint fails

#### Why are we doing this? How does it help us?
+ Time spent in code review talking about style is time wasted. Prettier takes care of JS formatting without anyone having to even think about it.
+ JavaScript is changing at an extraordinary rate, meaning it's easy to do things the old/ES5 way simply because you don't know a better way exists. By enforcing ESLint rules, we'll always be up-to-date on modern JS syntax, especially when it comes to Vue.js.

#### How should this be manually tested?
+ Install the pre-commit package: `brew install pre-commit`
+ Ensure you have a relatively up-to-date version of Node on your machine (> 8 ideally).
+ Do `pre-commit install` at the root of this directory.
+ Create a branch off this one (`git checkout -b foo-branch js-ci`)
+ Remove the semicolon from [this line](https://github.com/texastribune/donations/blob/master/static/js/src/utils/addNumberCommas.js#L2).
+ Try committing that change and confirm you get a failure message. The JS file should've been corrected to now have a semicolon. Do `git status` and verify that the semicolon addition didn't get automatically committed and is waiting to be added to the staging area. (I'm opting for this approach after reading [this](https://github.com/pre-commit/pre-commit/issues/747#issuecomment-386782080)).

#### How should this change be communicated to end users?
NA.

#### Are there any smells or added technical debt to note?
Our CircleCI builds will take 15-20 seconds longer now.

#### What are the relevant tickets?
https://3.basecamp.com/3098728/buckets/736178/todos/1457497379

#### Have you done the following, if applicable:
* [ ] Added automated tests? *(NA)*
